# XGBoost for modelling the amount of Code Smells
XGBoost is a state of the art tree model which is fit for a regression approach and has the advantage of being able to handle missing data.

In [1]:
import numpy as np
import pandas as pd
import os
import dataloader_functions.dataloader_fulldata_AmountCodeSmells as dataloader
from sklearn.model_selection import train_test_split, GridSearchCV, KFold
from xgboost import XGBRegressor
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
from joblib import dump, load

In [2]:
# define necessary variables for data setup
variable_list = [
    'CODE_SMELLS',
    'CLASSES',
    'FILES',
    'LINES',
    'NCLOC',
    'PACKAGE',
    'STATEMENTS',
    'FUNCTIONS',
    'COMMENT_LINES',
    'COMPLEXITY',
    'CLASS_COMPLEXITY',
    'FUNCTION_COMPLEXITY',
    'COGNITIVE_COMPLEXITY',
    'LINES_TO_COVER',
    'UNCOVERED_LINES',
    'DUPLICATED_LINES',
    'DUPLICATED_BLOCKS',
    'DUPLICATED_FILES',
    'COMMENT_LINES_DENSITY',
    'DUPLICATED_LINES_DENSITY'
]

label = "CODE_SMELLS"

current_dir = os.getcwd()
model_save_dir = os.path.join(current_dir, '..', '..', 'Data', 'Models', 'CodeSmells')

df = dataloader.load_df(current_dir)
df = dataloader.put_label_in_front(df, label)
df = dataloader.select_variables(df, variable_list)
df = dataloader.scale_predictors(df, label)
df

CODE_SMELLS   CLASSES     FILES     LINES     NCLOC   PACKAGE  \
0             17012  0.707998  0.114798  0.314222  0.571297  0.342736   
1             16987  0.707998  0.114798  0.310393  0.565516  0.342736   
2             16987  0.707998  0.114798  0.310393  0.565516  0.342736   
3             17013  0.707998  0.114798  0.314070  0.571095  0.342736   
4             17013  0.707998  0.114798  0.314053  0.571070  0.342736   
...             ...       ...       ...       ...       ...       ...   
140743          268 -0.768491 -0.779031 -0.738544 -0.683567 -0.772243   
140744          269 -0.768491 -0.779031 -0.738511 -0.683435 -0.772243   
140745          266 -0.768491 -0.779031 -0.738908 -0.683901 -0.772243   
140746          266 -0.768491 -0.779031 -0.738908 -0.683901 -0.772243   
140747          266 -0.768491 -0.779031 -0.738908 -0.683901 -0.772243   

        STATEMENTS  FUNCTIONS  COMMENT_LINES  COMPLEXITY  CLASS_COMPLEXITY  \
0         0.841017   0.593938      -0.346117    0.821715          1.442628   
1         0.841017   0.593938      -0.346196    0.821715          1.442628   
2         0.841017   0.593938      -0.346196    0.821715          1.442628   
3         0.841017   0.593938      -0.346196    0.821715          1.442628   
4         0.841017   0.593938      -0.346196    0.821715          1.442628   
...            ...        ...            ...         ...               ...   
140743   -0.666976  -0.664161      -0.819923   -0.633137         -0.830480   
140744   -0.666874  -0.664385      -0.820042   -0.633035         -0.830480   
140745   -0.667108  -0.664459      -0.820834   -0.633240         -0.830480   
140746   -0.667108  -0.664459      -0.820834   -0.633240         -0.830480   
140747   -0.667108  -0.664459      -0.820834   -0.633240         -0.830480   

        FUNCTION_COMPLEXITY  COGNITIVE_COMPLEXITY  LINES_TO_COVER  \
0                  1.588837              1.352548        0.702365   
1                  1.588837              1.352548        0.702365   
2                  1.588837              1.352548        0.702365   
3                  1.588837              1.352548        0.702365   
4                  1.588837              1.352548        0.702365   
...                     ...                   ...             ...   
140743            -0.629689             -0.670437       -0.724777   
140744            -0.629689             -0.670230       -0.724693   
140745            -0.629689             -0.670593       -0.725016   
140746            -0.629689             -0.670593       -0.725016   
140747            -0.629689             -0.670593       -0.725016   

        UNCOVERED_LINES  DUPLICATED_LINES  DUPLICATED_BLOCKS  \
0              0.702365          0.407604           0.610631   
1              0.702365          0.404699           0.609715   
2              0.702365          0.404699           0.609715   
3              0.702365          0.407604           0.610631   
4              0.702365          0.407604           0.610631   
...                 ...               ...                ...   
140743        -0.724777         -0.527839          -0.490534   
140744        -0.724693         -0.527839          -0.490534   
140745        -0.725016         -0.527839          -0.490534   
140746        -0.725016         -0.527839          -0.490534   
140747        -0.725016         -0.527839          -0.490534   

        DUPLICATED_FILES  COMMENT_LINES_DENSITY  DUPLICATED_LINES_DENSITY  
0               0.166980              -1.710559                  0.262477  
1               0.157664              -1.710559                  0.262477  
2               0.157664              -1.710559                  0.262477  
3               0.166980              -1.710559                  0.262477  
4               0.166980              -1.710559                  0.262477  
...                  ...                    ...                       ...  
140743         -0.792615               0.402842                 

## Train-Test-Split

In [3]:
# train-test-split
X = df.drop(columns = 'CODE_SMELLS')
y = df['CODE_SMELLS']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

## XGBoost model with tuned parameters
To ensure good model performance, a range of parameters is cross-validated with grid search. The model optimizes the MAE.

In [4]:
xgb_model = XGBRegressor(objective='reg:absoluteerror', random_state=42)

param_grid = {
    'n_estimators': [100, 200, 300, 400],
    'learning_rate': [0.1, 0.2, 0.5],
    'max_depth': [4, 5, 6, 7],
    'subsample': [0.8, 0.9],
    'colsample_bytree': [0.8, 0.9]
}

# 5-fold cross validation setup
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# grid search on param_grid, optimizing MAE
grid_search = GridSearchCV(
    estimator=xgb_model,
    param_grid=param_grid,
    scoring='neg_mean_absolute_error',
    cv=kf,
    verbose=2,
    n_jobs=-1,
)

grid_search.fit(X_train, y_train)

# extract best model
best_model = grid_search.best_estimator_
best_params = grid_search.best_params_

print("Best Parameters:", best_params)

# predictions for test data
y_pred = best_model.predict(X_test)

# evaluation
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)

print("XGBoost Model Evaluation (Tuned):")
print(f"Mean Squared Error: {mse}")
print(f"R-squared: {r2}")
print(f"Mean Absolute Error: {mae}")

filename_joblib = os.path.join(model_save_dir, "XGBoost", "XGBoost_tunedParams.joblib")
dump(best_model, filename_joblib)

Fitting 5 folds for each of 192 candidates, totalling 960 fits
Best Parameters: {'colsample_bytree': 0.8, 'learning_rate': 0.5, 'max_depth': 7, 'n_estimators': 400, 'subsample': 0.8}
XGBoost Model Evaluation (Tuned):
Mean Squared Error: 59443.22922451508
R-squared: 0.9995165395533762
Mean Absolute Error: 72.77299885213975


['C:\\Users\\carol\\Dropbox\\DataScience\\Semester4\\MasterProjectSonarQube\\Scripts\\Model\\..\\..\\Data\\Models\\CodeSmells\\XGBoost\\XGBoost_tunedParams.joblib']

## Result
By using XGBoost, a small MAE of 72.77 is achieved as testing performance. The MSE is low with 59443. XGBoost performs by far the best out of the tried model types.